src/IO/S3/copyS3File.cpp

```diff
@@ -406,14 +413,17 @@ namespace
             if (multipart_upload_aborted)
                 return; /// Already aborted.
 
-            auto tag = processUploadPartRequest(*task.req);
+            auto request = makeUploadPartRequest(task.part_number, task.part_offset, task.part_size);
+            auto tag = processUploadPartRequest(*request);
 
             std::lock_guard lock(bg_tasks_mutex); /// Protect bg_tasks from race
             task.tag = tag;
-            LOG_TRACE(log, "Writing part finished. Bucket: {}, Key: {}, Upload_id: {}, Etag: {}, Parts: {}", dest_bucket, dest_key, multipart_upload_id, task.tag, bg_tasks.size());
+            ++num_finished_parts;
+            LOG_TRACE(log, "Finished writing part #{}. Bucket: {}, Key: {}, Upload_id: {}, Etag: {}, Finished parts: {} of {}",
+                      task.part_number, dest_key, multipart_upload_id, task.tag, bg_tasks.size(), num_finished_parts, num_parts);
         }
 
-        virtual std::unique_ptr<Aws::AmazonWebServiceRequest> fillUploadPartRequest(size_t part_number, size_t part_offset, size_t part_size) = 0;
+        virtual std::unique_ptr<Aws::AmazonWebServiceRequest> makeUploadPartRequest(size_t part_number, size_t part_offset, size_t part_size) = 0;
```

Comment:
AFAIU previously this method was called in one thread from uploadPart() and with the fix it will be called from processUploadTask(*task) scheduled in a thread pool. Looks like it can be declared `const` and have a comment that derived class needs to take care about synchronization.

Reply:
done